source/common/network/listener_impl.cc

```diff
@@ -43,85 +41,92 @@ bool ListenerImpl::rejectCxOverGlobalLimit() {
-  // We work around this by using our own name for the socket in this case.
-  // Pass the 'v6only' parameter as true if the local_address is an IPv6 address. This has no effect
-  // if the socket is a v4 socket, but for v6 sockets this will create an IPv4 remote address if an
-  // IPv4 local_address was created from an IPv6 mapped IPv4 address.
-  const Address::InstanceConstSharedPtr& remote_address =
-      (remote_addr->sa_family == AF_UNIX)
-          ? io_handle->peerAddress()
-          : Address::addressFromSockAddr(*reinterpret_cast<const sockaddr_storage*>(remote_addr),
-                                         remote_addr_len,
-                                         local_address->ip()->version() == Address::IpVersion::v6);
-  listener->cb_.onAccept(
-      std::make_unique<AcceptedSocketImpl>(std::move(io_handle), local_address, remote_address));
 }
 
 void ListenerImpl::setupServerSocket(Event::DispatcherImpl& dispatcher, Socket& socket) {
-  listener_.reset(
-      evconnlistener_new(&dispatcher.base(), listenCallback, this, 0, -1, socket.ioHandle().fd()));
-
-  if (!listener_) {
-    throw CreateListenerException(
-        fmt::format("cannot listen on socket: {}", socket.localAddress()->asString()));
-  }
+#ifdef WIN32
+  auto trigger = Event::FileTriggerType::Level;
+#else
+  auto trigger = Event::FileTriggerType::Edge;
+#endif
+  socket.ioHandle().listen(128);
+  file_event_ = dispatcher.createFileEvent(
+      socket.ioHandle().fd(), [this](uint32_t events) -> void { onSocketEvent(events); }, trigger,
+      Event::FileReadyType::Read);
+
+  RELEASE_ASSERT(file_event_, "file event must be valid");
 
   if (!Network::Socket::applyOptions(socket.options(), socket,
                                      envoy::config::core::v3::SocketOption::STATE_LISTENING)) {
     throw CreateListenerException(fmt::format("cannot set post-listen socket option on socket: {}",
                                               socket.localAddress()->asString()));
   }
-
-  evconnlistener_set_error_cb(listener_.get(), errorCallback);
 }
 
 ListenerImpl::ListenerImpl(Event::DispatcherImpl& dispatcher, SocketSharedPtr socket,
                            ListenerCallbacks& cb, bool bind_to_port)
-    : BaseListenerImpl(dispatcher, std::move(socket)), cb_(cb), listener_(nullptr) {
+    : BaseListenerImpl(dispatcher, std::move(socket)), cb_(cb) {
   if (bind_to_port) {
     setupServerSocket(dispatcher, *socket_);
   }
 }
 
-void ListenerImpl::errorCallback(evconnlistener*, void*) {
-  // We should never get an error callback. This can happen if we run out of FDs or memory. In those
-  // cases just crash.
-  PANIC(fmt::format("listener accept failure: {}", errorDetails(errno)));
-}
-
-void ListenerImpl::enable() {
-  if (listener_.get()) {
-    evconnlistener_enable(listener_.get());
+ListenerImpl::~ListenerImpl() {
+  if (file_event_.get()) {
+    file_event_->setEnabled(0);
+    file_event_.reset();
```

Comment:
I'm fairly sure that you can just delete the file_event_, no need to setEnabled(0) first.  You should be able to remove this explicit destructor.

Reply:
Fair point. Let me run the whole test suite to see if it hits any problems.

Reply:
Out of the 687 tests, `//test/integration:h1_capture_fuzz_test` crashed at `source/common/http/conn_manager_impl.cc:639` on first suite run but always passed afterwards (if run as part of the suite or independently). Not sure what to make of it as it could be unrelated ...